dalm/eval/eval_rag.py

```diff
@@ -135,118 +133,72 @@ def run_generator_on_prompts(
-        return is_in_unique_list
-
-    unique_passage_dataset = processed_datasets.filter(is_passage_unique)
-
-    passage_to_id_dict = {i: p[args.passage_column_name] for i, p in enumerate(unique_passage_dataset)}
+def eval_generator_on_batch(
+    model: PreTrainedModel,
+    tokenizer: PreTrainedTokenizer,
+    queries: List[str],
+    passages: List[str],
+    query_batch_size: int,
+    queries_for_gen_eval: list[str],
+    max_length: int,
+) -> tuple[list[str], list[str]]:
+    generated_answers_for_eval = []
+    for _query, search_result_passage in zip(queries, passages, strict=True):
+        # This query comes without the answer
+        query = f"#query# {_query} #passage# {search_result_passage} #answer# "
+        queries_for_gen_eval.append(query)
+        # Generate answers in batch
+        if len(queries_for_gen_eval) >= query_batch_size:
+            batch_answers = run_generator_on_prompts(model, tokenizer, queries_for_gen_eval, max_length=max_length)
+            generated_answers_for_eval.extend(batch_answers)
+            queries_for_gen_eval.clear()
 
-    unique_passage_dataloader = DataLoader(
-        unique_passage_dataset,
-        shuffle=False,
-        collate_fn=default_data_collator,
-        batch_size=args.test_batch_size,
-        pin_memory=True,
+    return queries_for_gen_eval, generated_answers_for_eval
+
+
+def evaluate_rag(
+    dataset_or_path: str | Dataset,
+    retriever_name_or_path: str,
+    generator_name_or_path: str,
+    retriever_peft_model_path: str,
+    generator_peft_model_path: str,
+    passage_column_name: str,
+    query_column_name: str,
+    answer_column_name: str,
+    embed_dim: int,
+    max_length: int,
+    test_batch_size: int = 8,
+    query_batch_size: int = 16,
+    device: str = "cuda",
+    torch_dtype: Literal["float16", "bfloat16"] = "float16",
+    top_k: int = 10,
+    evaluate_generator: bool = True,
+) -> None:
+    """Runs rag evaluation. See `dalm eval-rag --help for details on params"""
+    test_dataset = load_dataset(dataset_or_path)
+    selected_torch_dtype: Final[torch.dtype] = torch.float16 if torch_dtype == "float16" else torch.bfloat16
+    # rag retriever and the generator (don't load new peft layers no need)
+    rag_model = AutoModelForRagE2E(retriever_name_or_path, generator_name_or_path, get_peft=False, use_bnb=False)
+
+    processed_datasets = preprocess_dataset(
+        test_dataset, rag_model.retriever_tokenizer, query_column_name, passage_column_name, max_length
     )
-
     # peft config and wrapping
-    rag_model.attach_pre_trained_peft_layers(
-        args.retriever_peft_model_path, args.generator_peft_model_path, args.device
+    rag_model.attach_pre_trained_peft_layers(retriever_peft_model_path, generator_peft_model_path, device)
+    unique_passage_dataset, passage_embeddings_array = get_passage_embeddings(
+        processed_datasets,
+        passage_column_name,
+        rag_model.retrieval_forward,
```

Comment:
Is this okay to pass a function? I am ok with it. Is there a better way?

Reply:
it's pretty normal to pass callables `typing.Callable` around :)